uniqueptr.h

```diff
@@ -0,0 +1,113 @@
+#ifndef UNIQUEPTR_H
+#define UNIQUEPTR_H
+
+#include <memory>
+
+namespace smart_home
+{
+
+template <class T, class Deleter = std::default_delete<T>> class UniquePtr
+{
+public:
+    UniquePtr(T* ptr = nullptr, Deleter&& deleter = Deleter()) noexcept : ptr_{ptr}, deleter_{deleter}{}
+    UniquePtr(UniquePtr &&ptr): ptr_{std::move(ptr)}{}
+    UniquePtr(UniquePtr &ptr) = delete;
+    UniquePtr& operator=(const UniquePtr& ptr) = delete;
+    UniquePtr& operator=(const UniquePtr&& ptr) noexcept
+    {
+        if (this != ptr)
+        {
+            ptr_.reset(ptr.release());
+            this.deleter_ = std::move(ptr->deleter_);
+        }
+
+        return *this;
+    }
+
+    //Modifiers
+    T* release() noexcept
+    {
+        T* released = ptr_;
+        ptr_ = nullptr;
+        return released;
+    }
+
+    void reset( T* ptr = new T()) noexcept
+    {
+        T* oldPtr = this->ptr_;
+        this->ptr_= ptr;
+
+        if (oldPtr)
+        {
+            deleter_(oldPtr);
+        }    
+    }
+
+    void swap( UniquePtr& other ) noexcept
+    {
+        if(ptr_ != other.ptr_)
+        {
+            T* auxPtr = this->release();
+            ptr_ = other.release();
+            other.ptr_ = auxPtr;
+
+            auxPtr = nullptr;
+            delete auxPtr;
```

Comment:
this part of code doesn't look just right to me, you're deleting a nullptr? why?